File: uploader/src/main/java/net/hearthstats/Card.java
package net.hearthstats;

import lombok.Value;
import lombok.experimental.Accessors;
import lombok.experimental.Builder;

@Value
@Builder
@Accessors(fluent = true)
public class Card implements Comparable<Card> {
	private int id;
	private String name;
	private int count;
	private int cost;

	public String fileName() {
		return String.format("%s.png", name.replaceAll("[^a-zA-Z]", "-")
				.toLowerCase());
	}

	public String url() {
		return String.format(
				"https://s3-us-west-2.amazonaws.com/hearthstats/cards/%s",
				fileName());
	}

	@Override
	public int compareTo(Card c) {
		int costs = Integer.compare(cost, c.cost);
		if (costs != 0)
			return costs;
		int counts = Integer.compare(count, c.count);
		if (counts != 0)
			return counts;
		return name.compareTo(c.name);
	}
}

File: uploader/src/main/scala/net/hearthstats/Card.scala
package net.hearthstats

import java.io.File
import java.net.MalformedURLException
import java.net.URL
import Card._
import net.hearthstats.util.TranslationCard

object Card {
  val LEGENDARY = 5
  private lazy val imageCacheFolder: String = Config.getImageCacheFolder
}

case class Card(
  id: Int,
  originalName: String,
  count: Int = 0,
  cost: Int = 0,
  rarity: Int = 0,
  collectible: Boolean = true)
  extends Comparable[Card] {

  var name = if (TranslationCard.hasKey(id.toString) == false) originalName else TranslationCard.t(id.toString)

  def isLegendary: Boolean = rarity == LEGENDARY

  def fileName: String =
    String.format("%s.png", originalName.replaceAll("[ :']+", "-").replaceAll("""[^a-zA-Z0-9\-]""", "").toLowerCase)

  def localURL: URL = localFile.toURI.toURL

  def localFile: File = new File(imageCacheFolder, fileName)

  def url: String =
    String.format("https://s3-us-west-2.amazonaws.com/hearthstats/cards/%s", fileName)

  override def compareTo(c: Card): Int = {
    val costs = Integer.compare(cost, c.cost)
    if (costs != 0) return costs
    val counts = Integer.compare(count, c.count)
    if (counts != 0) return counts
    originalName.compareTo(c.originalName)
  }
}
